pkg/executor/builder.go

```diff
@@ -2972,13 +2972,18 @@ func markChildrenUsedCols(outputCols []*expression.Column, childSchemas ...*expr
 
 func (*executorBuilder) corColInDistPlan(plans []plannercore.PhysicalPlan) bool {
 	for _, p := range plans {
-		x, ok := p.(*plannercore.PhysicalSelection)
-		if !ok {
-			continue
-		}
-		for _, cond := range x.Conditions {
-			if len(expression.ExtractCorColumns(cond)) > 0 {
-				return true
+		switch x := p.(type) {
```

Comment:
Why mpp plan works fine?

Reply:
mpp returns a `MPPGather` when `buildTableReader `

https://github.com/pingcap/tidb/blob/cf23666766d03ec9c2551465fbb92e0f9bba25a2/pkg/executor/builder.go#L3414-L3435

MPPGather will always call `ConstructTreeBasedDistExec` when `ConstructDAGReq`

https://github.com/pingcap/tidb/blob/cf23666766d03ec9c2551465fbb92e0f9bba25a2/pkg/executor/internal/builder/builder_utils.go#L46-L59

For `TableReaderExecutor`, only when `e.corColInFilter = true` will call `ConstructTreeBasedDistExec`

https://github.com/pingcap/tidb/blob/cf23666766d03ec9c2551465fbb92e0f9bba25a2/pkg/executor/table_reader.go#L142-L161